File: basex-api/src/main/c#/BaseXClient.cs
/*
 * Language Binding for BaseX.
 * Works with BaseX 7.0 and later
 *
 * Documentation: https://docs.basex.org/wiki/Clients
 *
 * (C) BaseX Team 2005-12, BSD License
 */
using System;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.IO;
using System.Threading;
using System.Collections;
using System.Collections.Generic;

namespace BaseXClient
{
  public class Session
  {
    private byte[] cache = new byte[4096];
    public NetworkStream stream;
    private TcpClient socket;
    private string info = "";
    private int bpos;
    private int bsize;

    public Session(string host, int port, string username, string pw)
    {
      socket = new TcpClient(host, port);
      stream = socket.GetStream();
      string[] response = Receive().Split(':');

      string nonce;
      string code;
      if (response.Length > 1)
      {
          code = username + ":" + response[0] + ":" + pw;
          nonce = response[1];
      }
      else
      {
          code = pw;
          nonce = response[0];
      }

      Send(username);
      Send(MD5(MD5(code) + nonce));
      if (stream.ReadByte() != 0)
      {
          throw new IOException("Access denied.");
      }
    }

    public void Execute(string com, Stream ms)
    {
      Send(com);
      Init();
      Receive(ms);
      info = Receive();
      if(!Ok())
      {
        throw new IOException(info);
      }
    }

    public String Execute(string com)
    {
      MemoryStream ms = new MemoryStream();
      Execute(com, ms);
      return System.Text.Encoding.UTF8.GetString(ms.ToArray());
    }

    public Query Query(string q)
    {
      return new Query(this, q);
    }

    public void Create(string name, Stream s)
    {
      stream.WriteByte(8);
      Send(name);
      Send(s);
    }
    
    public void Add(string path, Stream s)
    {
      stream.WriteByte(9);
      Send(path);
      Send(s);
    }
    
    public void Replace(string path, Stream s)
    {
      stream.WriteByte(12);
      Send(path);
      Send(s);
    }
    
    public void Store(string path, Stream s)
    {
      stream.WriteByte(13);
      Send(path);
      Send(s);
    }

    public string Info
    {
      get
      {
        return info;
      }
    }

    public void Close()
    {
      Send("exit");
      socket.Close();
    }

    private void Init()
    {
      bpos = 0;
      bsize = 0;
    }

    public byte Read()
    {
      if (bpos == bsize)
      {
        bsize = stream.Read(cache, 0, 4096);
        bpos = 0;
      }
      return cache[bpos++];
    }

    private void Receive(Stream ms)
    {
      while (true)
      {
        byte b = Read();
        if (b == 0) break;
        // read next byte if 0xFF is received
        ms.WriteByte(b == 0xFF ? Read() : b);
      }
    }

    public string Receive()
    {
      MemoryStream ms = new MemoryStream();
      Receive(ms);
      return System.Text.Encoding.UTF8.GetString(ms.ToArray());
    }

    public void Send(string message)
    {
      byte[] msg = System.Text.Encoding.UTF8.GetBytes(message);
      stream.Write(msg, 0, msg.Length);
      stream.WriteByte(0);
    }

    private void Send(Stream s)
    {
      while (true)
      {
          int t = s.ReadByte();
          if (t == -1) break;
          if (t == 0x00 || t == 0xFF) stream.WriteByte(Convert.ToByte(0xFF));
          stream.WriteByte(Convert.ToByte(t));
      }
      stream.WriteByte(0);
      info = Receive();
      if(!Ok())
      {
        throw new IOException(info);
      }
    }
    

    public bool Ok()
    {
      return Read() == 0;
    }

    private string MD5(string input)
    {
      MD5CryptoServiceProvider MD5 = new MD5CryptoServiceProvider();
      byte[] hash = MD5.ComputeHash(Encoding.UTF8.GetBytes(input));

      StringBuilder sb = new StringBuilder();
      foreach (byte h in hash)
      {
        sb.Append(h.ToString("x2"));
      }
      return sb.ToString();
    }
  }

  public class Query
  {
    private Session session;
    private string id;
    private ArrayList cache;
    private int pos;

    public Query(Session s, string query)
    {
      session = s;
      id = Exec(0, query);
    }

    public void Bind(string name, string value)
    {
      Bind(name, value, "");
    }

    public void Bind(string name, string value, string type)
    {
      cache = null;
      Exec(3, id + '\0' + name + '\0' + value + '\0' + type);
    }

    public void Context(string value)
    {
      Context(value, "");
    }

    public void Context(string value, string type)
    {
      cache = null;
      Exec(14, id + '\0' + value + '\0' + type);
    }

    public bool More()
    {
      if(cache == null) 
      {
        session.stream.WriteByte(4);
        session.Send(id);
        cache = new ArrayList();
        while (session.Read() > 0)
        {
          cache.Add(session.Receive());
        }
        if(!session.Ok())
        {
          throw new IOException(session.Receive());
        }
        pos = 0;
      }
      if(pos < cache.Count) return true;
      cache = null;
      return false;
    }

    public string Next()
    {
      if(More()) 
      {
        return cache[pos++] as string;
      }
      else
      {
        return null;
      }
    }

    public string Execute()
    {
      return Exec(5, id);
    }

    public string Info()
    {
      return Exec(6, id);
    }

    public string Options()
    {
      return Exec(7, id);
    }

    public void Close()
    {
      Exec(2, id);
    }

    private string Exec(byte cmd, string arg)
    {
      session.stream.WriteByte(cmd);
      session.Send(arg);
      string s = session.Receive();
    bool ok = session.Ok();
      if(!ok)
      {
        throw new IOException(session.Receive());
      }
      return s;
    }
  }
}

File: basex-api/src/main/vb/BaseXClient.vb
' Visual Basic client for BaseX.
' Works with BaseX 7.x (but not with BaseX 8.0 and later)
' Does not support all bindings yet; your extensions are welcome.
'
' Documentation: http://docs.basex.org/wiki/Clients
'
' (C) BaseX Team 2005-12, BSD License

Imports System
Imports System.Net.Sockets
Imports System.Security.Cryptography
Imports System.Text
Imports System.Collections.Generic
Imports System.IO

Module BaseXClient
  Class Session	
    Private cache As Byte() = New Byte(4096) {}
    Public stream As NetworkStream
    Private socket As TcpClient
    Private m_info As String = ""
    Private bpos As Integer
    Private bsize As Integer
    
    Public Sub New(host As String, port As Integer, username As String, pw As String)
      socket = New TcpClient(host, port)
      stream = socket.GetStream()
      Dim ts As String = Receive()
      Send(username)
      Dim tmp As String = MD5StringHash(pw)
      Send(MD5StringHash(tmp & ts))
      If stream.ReadByte() <> 0 Then
        Throw New IOException("Access denied.")
      End If	
    End Sub
    
    Public Sub Execute(com As String, ms As Stream)
      Send(com)
      Init()
      Receive(ms)
      m_info = Receive()
      If Not Ok() Then
        Throw New IOException(m_info)
      End If
    End Sub

    Public Function Execute(com As String) As [String]
      Dim ms As New MemoryStream()
      Execute(com, ms)
      Return System.Text.Encoding.UTF8.GetString(ms.ToArray())
    End Function
    
    Public Sub Create(name As String, ms As Stream)
      stream.WriteByte(8)
      Send(name)
      While True:
      	Dim t As Integer = ms.ReadByte()
      	If t = -1 Then
      		Exit While
      	End If
      	stream.WriteByte(Convert.ToByte(t))
      End While
      stream.WriteByte(0)
      m_info = Receive()
      If Not Ok() Then
        Throw New IOException(m_info)
      End If
    End Sub
    
    Public Sub Create(name As String, target As String, ms As Stream)
      stream.WriteByte(9)
      Send(name)
      Send(target)
      While True:
      	Dim t As Integer = ms.ReadByte()
      	If t = -1 Then
      		Exit While
      	End If
      	stream.WriteByte(Convert.ToByte(t))
      End While
      stream.WriteByte(0)
      m_info = Receive()
      If Not Ok() Then
        Throw New IOException(m_info)
      End If
    End Sub

    Public Function Query(q As String) As Query
      Return New Query(Me, q)
    End Function

    Public ReadOnly Property Info() As String
      Get
        Return m_info
      End Get
    End Property

    Public Sub Close()
      Send("exit")
      socket.Close()
    End Sub

    ' Initializes the byte transfer. 
    Private Sub Init()
      bpos = 0
      bsize = 0
    End Sub

    ' Returns a single byte from the socket. 
    Private Function Read() As Byte
      If bpos = bsize Then
        bsize = stream.Read(cache, 0, 4096)
        bpos = 0
      End If
      Dim b as Byte = cache(bpos)
      bpos += 1
      Return b
    End Function

    ' Receives a string from the socket. 
    Private Sub Receive(ms As Stream)
      While True
        Dim b As Byte = Read()
        If b = 0 Then
          Exit While
        End If
        ms.WriteByte(b)
      End While
    End Sub

    ' Receives a string from the socket. 
    Public Function Receive() As String
      Dim ms As New MemoryStream()
      Receive(ms)
      Return System.Text.Encoding.UTF8.GetString(ms.ToArray())
    End Function

    ' Sends strings to server. 
    Public Sub Send(message As String)
      Dim msg As Byte() = System.Text.Encoding.UTF8.GetBytes(message)
      stream.Write(msg, 0, msg.Length)
      stream.WriteByte(0)
    End Sub

    ' Returns success check. 
    Public Function Ok() As Boolean
      Return Read() = 0
    End Function

    ' Returns the md5 hash of a string. 
  Private Function MD5StringHash(ByVal strString As String) As String
    Dim MD5 As New MD5CryptoServiceProvider
    Dim Data As Byte()
    Dim Result As Byte()
    Dim Res As String = ""
    Dim Tmp As String = ""

    Data = Encoding.ASCII.GetBytes(strString)
    Result = MD5.ComputeHash(Data)
    For i As Integer = 0 To Result.Length - 1
        Tmp = Hex(Result(i))
        If Len(Tmp) = 1 Then Tmp = "0" & Tmp
        Res += Tmp
    Next
    Return Res.ToLower
  End Function
End Class

  Class Query
    Private session As Session
    Private id As String

    Public Sub New(s As Session, query As String)
      session = s
      id = Exec(0, query)
    End Sub
      
    Public Sub Bind(name As String, value As String)
      Bind(name, value, "")
    End Sub

    Public Sub Bind(name As String, value As String, type As String)
      session.stream.WriteByte(3)
      session.Send(id)
      session.Send(name)
      session.Send(value)
      session.Send(type)
      Dim Res As String = session.Receive()
      If Not session.Ok() Then
        Throw New IOException(session.Receive())
      End If  
    End Sub

    Public Sub Context(value As String)
      Context(value, "")
    End Sub

    Public Sub Context(value As String, type As String)
      session.stream.WriteByte(14)
      session.Send(id)
      session.Send(value)
      session.Send(type)
      Dim Res As String = session.Receive()
      If Not session.Ok() Then
        Throw New IOException(session.Receive())
      End If  
    End Sub

    Public Function Execute() As String
      Return Exec(5, id)
    End Function

    Public Function Info() As String
      Return Exec(6, id)
    End Function

    Public Function Close()
      Exec(2, id)
    End Function
    
    Public Function Exec(cmd As Integer, arg As String) As String
      session.stream.WriteByte(cmd)
      session.Send(arg)
      Dim Res As String = session.Receive()
      If Not session.Ok() Then
        Throw New IOException(session.Receive())
      End If
      Return Res
    End Function
          
  End Class
End Module
